styx-common/src/main/java/com/spotify/styx/storage/BigtableStorage.java

```diff
@@ -127,9 +134,17 @@ void writeEvent(SequenceEvent sequenceEvent) throws IOException {
           final int lastHash = key.lastIndexOf('#');
 
           final WorkflowInstance wfi = WorkflowInstance.parseKey(key.substring(0, lastHash));
-          workflowInstancesSet.add(wfi);
-          if (workflowInstancesSet.size() == limit) {
-            break;
+          final boolean added = workflowInstancesSet.add(wfi);
+          if (added) {
+            if (tail) {
+              tailInstances.addLast(wfi);
```

Comment:
Can this logic be simplified using `EvictingQueue`? So we don't need to explicitly remove the head.

Reply:
Then we don't know which wfi to remove from `workflowInstancesSet`. Sure, we could refactor this even more to avoid the need for `workflowInstancesSet` but not sure it's worth it.